pkg/awsiamauth/reconciler/reconciler.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/aws/eks-anywhere/pkg/crypto"
+)
+
+// RemoteClientRegistry defines methods for remote cluster controller clients.
+type RemoteClientRegistry interface {
+	GetClient(ctx context.Context, cluster client.ObjectKey) (client.Client, error)
+}
+
+// Reconciler allows to reconcile AWSIamConfig.
+type Reconciler struct {
+	certgen              crypto.CertificateGenerator
+	templateBuilder      *awsiamauth.TemplateBuilder
+	generateUUID         UUIDGenerator
+	client               client.Client
+	remoteClientRegistry RemoteClientRegistry
+}
+
+// UUIDGenerator generates a new UUID.
+type UUIDGenerator func() uuid.UUID
+
+// New returns a new Reconciler.
+func New(certgen crypto.CertificateGenerator, generateUUID UUIDGenerator, client client.Client, remoteClientRegistry RemoteClientRegistry) *Reconciler {
+	return &Reconciler{
+		certgen:              certgen,
+		templateBuilder:      &awsiamauth.TemplateBuilder{},
+		generateUUID:         generateUUID,
+		client:               client,
+		remoteClientRegistry: remoteClientRegistry,
+	}
+}
+
+// EnsureCASecret ensures the AWS IAM Authenticator secret is present.
+// It uses a controller.Result to indicate when requeues are needed.
+func (r *Reconciler) EnsureCASecret(ctx context.Context, logger logr.Logger, cluster *anywherev1.Cluster) (controller.Result, error) {
+	clusterName := cluster.Name
+	s := &corev1.Secret{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: awsiamauth.CASecretName(clusterName), Namespace: constants.EksaSystemNamespace}, s)
+	if apierrors.IsNotFound(err) {
+		logger.Info("Creating aws-iam-authenticator CA secret")
+		return r.createCASecret(ctx, cluster)
+	}
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("fetching secret %s: %v", awsiamauth.CASecretName(clusterName), err)
+	}
+
+	logger.Info("aws-iam-authenticator CA secret found. Skipping secret create.", "Name", awsiamauth.CASecretName(clusterName))
+	return controller.Result{}, nil
+}
+
+func (r *Reconciler) createCASecret(ctx context.Context, cluster *anywherev1.Cluster) (controller.Result, error) {
+	yaml, err := r.templateBuilder.GenerateCertKeyPairSecret(r.certgen, cluster.Name)
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("generating aws-iam-authenticator ca secret: %v", err)
+	}
+
+	return controller.Result{}, serverside.ReconcileYaml(ctx, r.client, yaml)
+}
+
+// Reconcile takes the AWS IAM Authenticator installation to the desired state defined in AWSIAMConfig.
+// It uses a controller.Result to indicate when requeues are needed.
+// Intended to be used in a kubernetes controller.
+func (r *Reconciler) Reconcile(ctx context.Context, logger logr.Logger, cluster *anywherev1.Cluster) (controller.Result, error) {
+	clusterSpec, err := anywhereCluster.BuildSpec(ctx, clientutil.NewKubeClient(r.client), cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	result, err := clusters.CheckControlPlaneReady(ctx, r.client, logger, cluster)
```

Comment:
this was intended to be used as a cluster reconciler phase. The interface (using Result) doesn't fit this code very well. We should probably some of move the logic inside `CheckControlPlaneReady` to a shared method that just returns true or false.

I'm ok doing this as a follow up.

Reply:
Makes sense. 
I will strip out the logic and fetch the capi object. Since I am fetching it again for ownerRef setting, I can prevent that and fetch it once.